src/Database/LSMTree/Model/Normal.hs

```diff
@@ -0,0 +1,258 @@
+    convertRange (FromToExcluding lb ub) = (Map.Range.Bound (serialise lb) Map.Range.Inclusive, Map.Range.Bound (serialise ub) Map.Range.Exclusive)
+    convertRange (FromToIncluding lb ub) = (Map.Range.Bound (serialise lb) Map.Range.Inclusive, Map.Range.Bound (serialise ub) Map.Range.Inclusive)
+
+-- | Normal tables support insert and delete operations.
+--
+-- An __update__ is a term that groups all types of table-manipulating
+-- operations, like inserts and deletes.
+data Update v blob =
+    Insert !v !(Maybe blob)
+  | Delete
+  deriving (Eq, Show)
+
+-- | Perform a mixed batch of inserts and deletes.
+--
+-- Updates can be performed concurrently from multiple Haskell threads.
+updates :: forall k v blob.
+     ( SomeSerialisationConstraint k
+     , SomeSerialisationConstraint v
+     , SomeSerialisationConstraint blob
+     )
+  => [(k, Update v blob)]
+  -> Table k v blob
+  -> Table k v blob
+updates ups tbl0 = foldl' update tbl0 ups where
+    update :: Table k v blob -> (k, Update v blob) -> Table k v blob
+    -- Note: blobs are not collected. Do we need to ref-count them?
+    update tbl (k, Delete) = tbl { _values = Map.delete (serialise k) (_values tbl) }
+    update tbl (k, Insert v Nothing) = tbl { _values = Map.insert (serialise k) (serialise v, Nothing) (_values tbl) }
+    update tbl (k, Insert v (Just blob)) = tbl
+        { _values = Map.insert (serialise k) (serialise v, Just (mkBlobRef blob)) (_values tbl)
+        }
+
+-- | Perform a batch of inserts.
+--
+-- Inserts can be performed concurrently from multiple Haskell threads.
+inserts ::
+     ( SomeSerialisationConstraint k
+     , SomeSerialisationConstraint v
+     , SomeSerialisationConstraint blob
+     )
+  => [(k, v, Maybe blob)]
+  -> Table k v blob
+  -> Table k v blob
+inserts = updates . fmap (\(k, v, blob) -> (k, Insert v blob))
+
+-- | Perform a batch of deletes.
+--
+-- Deletes can be performed concurrently from multiple Haskell threads.
+deletes ::
+     ( SomeSerialisationConstraint k
+     , SomeSerialisationConstraint v
+     , SomeSerialisationConstraint blob
+     )
+  => [k]
+  -> Table k v blob
+  -> Table k v blob
+deletes = updates . fmap (,Delete)
+
+-- | A reference to an on-disk blob.
+--
+-- The blob can be retrieved based on the reference.
+--
+-- Blob comes from the acronym __Binary Large OBject (BLOB)__ and in many
+-- database implementations refers to binary data that is larger than usual
+-- values and is handled specially. In our context we will allow optionally a
+-- blob associated with each value in the table.
+data BlobRef blob = BlobRef
+    !BS.ByteString  -- ^ digest
+    !BS.ByteString  -- ^ actual contents
+  deriving (Show)
+
+type role BlobRef nominal
+
+mkBlobRef :: SomeSerialisationConstraint blob => blob -> BlobRef blob
+mkBlobRef blob =  BlobRef (SHA256.hash bs) bs
+  where
+    !bs = serialise blob
+
+coerceBlobRef :: BlobRef blob -> BlobRef blob'
+coerceBlobRef (BlobRef d b) = BlobRef d b
```

Comment:
Only minor, since it's an internal function: this coercion makes sense for the `Table`'s show instance, but otherwise would probably break stuff. Maybe add it as a local def?

Reply:
Module has export list.